menpofit/dpm/base.py

```diff
@@ -0,0 +1,1062 @@
+        overlap: `double`
+            The minimum overlap ratio in latent positive search
+
+        Returns
+        -------
+        model: `DPM`
+            The model improved by the given positive and negative examples.
+        """
+        # the size of vectorized model
+        length = model.sparse_length()
+        # Maximum number of model that fitted into the cache
+        nmax = round(maxsize*0.25*10**9/length)
+        qp = Qp(model, length, nmax, c, wpos)
+        for t in range(iters):
+            # get positive examples using latent fitting. Found examples are saved to qp.x
+            model.delta = self._poslatent(t, model, qp, pos, overlap)
+            if model.delta < 0.001:  # terminate if the score doesn't change much
+                break
+            qp.svfix = range(qp.n)  # fix positive examples as permanent support vectors
+            qp.sv[qp.svfix] = True
+            qp.prune()
+            qp.opt(0.5)
+            model = qp.vec2model(model)
+            model.interval = 4
+
+            for i, neg in enumerate(negs):
+                print 'iter:', t, 'neg:', i, '/', np.size(negs)
+                im = mio.import_image(neg['im'])
+                box, model = DPMFitter.fit_with_bb(im,  model, -1, [], 0, i, -1, qp)
+                if np.sum(qp.sv) == nmax:
+                    break
+
+            qp.opt()
+            model = qp.vec2model(model)
+            root_scores = np.sort(qp.score_pos())
+            # compute minimum score on positive example (with raw, unscaled features)
+            model.thresh = root_scores[np.ceil[np.size(root_scores)*0.05]]
+            model.interval = 10
+            model.lb = qp.lb
+            model.ub = qp.ub
+
+        return model
+
+    def _poslatent(self, t, model, qp, poses, overlap):
+        r"""
+        Get positive examples using latent fitting. A true bbox is used to check the overlapping with the fitting
+        results. The positive examples are saved into qp by calling qp.write
+
+        Parameters
+        ----------
+        t: `int`
+            The number of iteration(s)
+        model: `DPM`
+            The model which will be improved using the given positive and negative examples
+        qp: `Qp`
+            The dual coordinate solver which created by the given model
+        poses: `list`
+            A list of positive examples(dictionary) which contain associated image part and bounding boxes
+        overlap: `double`
+            The minimum overlap ratio in latent positive search
+
+        Returns
+        -------
+        delta: `double`
+            The different of scores before and after the model is trained(improved)
+        """
+        num_pos = np.size(poses)
+        model.interval = 5
+        num_positives = np.zeros(np.size(model.components, ), dtype=int)
+        score0 = qp.score_pos()
+        qp.n = 0
+        old_w = qp.w
+        qp = qp.model2qp(model)
+        assert(scipy.linalg.norm(old_w - qp.w) < 10**-5)
```

Comment:
These assertions are fine - but they will raise an ``AssertionError`` if violated - would be nice to have an error message telling the user what was violated. Maybe raise a `ValueError` or a new `TrainingError` or something.